src/providers/clojure.rs

```diff
@@ -20,21 +22,38 @@ impl Provider for ClojureProvider {
         Ok(app.includes_file("project.clj"))
     }
 
-    fn setup(&self, _app: &App, _env: &Environment) -> Result<Option<SetupPhase>> {
+    fn install(&self, _app: &App, _env: &Environment) -> Result<Option<InstallPhase>> {
+        let mut install_phase = InstallPhase::new("".to_string());
+        install_phase.add_cache_directory(LEIN_CACHE_DIR.to_string());
+        Ok(Some(install_phase))
+    }
+
+    fn setup(&self, app: &App, env: &Environment) -> Result<Option<SetupPhase>> {
         Ok(Some(SetupPhase::new(vec![
             Pkg::new("leiningen"),
-            Pkg::new("jdk8"),
+            ClojureProvider::get_nix_jdk_package(app, env)?,
         ])))
     }
 
-    fn build(&self, _app: &App, _env: &Environment) -> Result<Option<BuildPhase>> {
-        Ok(Some(BuildPhase::new("lein uberjar".to_string())))
+    fn build(&self, app: &App, _env: &Environment) -> Result<Option<BuildPhase>> {
+        let has_lein_ring_plugin = app
+            .read_file("project.clj")?
+            .to_lowercase()
+            .contains("[lein-ring ");
+
+        let build_cmd = match has_lein_ring_plugin {
+            true => "lein ring uberjar",
+            false => "lein uberjar",
+        };
+
+        Ok(Some(BuildPhase::new(build_cmd.to_string())))
     }
 
     fn start(&self, _app: &App, _env: &Environment) -> Result<Option<StartPhase>> {
-        Ok(Some(StartPhase::new(
-            "java $JAVA_OPTS -jar target/uberjar/*standalone.jar".to_string(),
-        )))
+        // based on project config, uberjar can be created under ./target/uberjar or ./target, this ensure file will be found on the same place for different project configs
+        let start_cmd= "mv /app/target/uberjar/*standalone.jar /app/target/*standalone.jar; java $JAVA_OPTS -jar /app/target/*standalone.jar";
```

Comment:
Can the move happen during the build phase instead? To keep it consistent with the other providers